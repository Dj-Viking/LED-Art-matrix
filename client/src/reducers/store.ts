import { configureStore } from "@reduxjs/toolkit";
import { midiSlice } from "./midiSlice";
import { loggedInSlice } from "./loggedInSlice";
import { ledSlice } from "./ledSlice";
import { modalSlice } from "./modalSlice";
import { presetButtonsListSlice } from "./presetButtonListSlice";

export const toolkitStore = configureStore({
    reducer: {
        midiState: midiSlice.reducer,
        loggedInState: loggedInSlice.reducer,
        ledState: ledSlice.reducer,
        modalState: modalSlice.reducer,
        presetButtonsListState: presetButtonsListSlice.reducer,
    },
});

export type ToolkitRootState = ReturnType<typeof toolkitStore.getState>;

export type ToolkitDispatch = typeof toolkitStore.dispatch;
