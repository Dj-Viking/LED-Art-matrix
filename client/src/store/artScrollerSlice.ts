import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { IArtScrollerState, IGif } from "../types";
import { produce } from "immer";

const initialState: IArtScrollerState = {
    figureOn: false,
    gifs: [],
    listName: "",
    slider: {
        animDuration: "30",
        circleWidth: "30",
        hPos: "33",
        vertPos: "111",
        invert: "0",
    },
};

export const artScrollerSlice = createSlice({
    name: "artScrollerSlice",
    initialState,
    reducers: {
        setListName: (state: IArtScrollerState, action: PayloadAction<string>) => {
            return produce(state, () => {
                state.listName = action.payload;
            });
        },
        setGifs: (state: IArtScrollerState, action: PayloadAction<IGif[]>) => {
            return produce(state, () => {
                state.gifs = action.payload;
            });
        },
        setFigureOn: (state: IArtScrollerState, action: PayloadAction<boolean>) => {
            return produce(state, () => {
                state.figureOn = action.payload;
            });
        },
        setSlider: (
            state: IArtScrollerState,
            action: PayloadAction<{ control: keyof IArtScrollerState["slider"]; value: string }>
        ) => {
            return produce(state, () => {
                const { control, value } = action.payload;
                state.slider[control] = value;
            });
        },
    },
});

export const artScrollerActions = {
    ...artScrollerSlice.actions,
};
