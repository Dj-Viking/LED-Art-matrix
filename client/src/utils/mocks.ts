// import { rest } from "msw";
// import { setupServer } from "msw/node";

export const LOGIN_MOCK_PAYLOAD_USERNAME = {
  emailOrUsername: "i exist",
  password: "believe it"
};

export const LOGIN_MOCK_PAYLOAD_EMAIL = {
  emailOrUsername: "iexist@exist.com",
  password: "believe it"
};

export const LOGIN_MOCK_RESPONSE = {
  _id: "kdjfkdfj",
  token: "token man",
  defaultPreset: "waves"
};

export const LOGIN_MOCK_ERROR_CODE = {
  status: 400
};

export const SIGNUP_MOCK_NO_TOKEN = {
  token: void 0,
  _id: "weeeeeeeee"
};

export const SIGNUP_MOCK_PAYLOAD_SAME_USER = {
  username: "test user",
  email: "testemail@email.com",
  password: "test password",
};
export const SIGNUP_MOCK_PAYLOAD = {
  username: "test user" + Date.now(),
  email: "testemail" + Date.now() + "@email.com",
  password: "test password" + Date.now(),
};

export const SIGNUP_MOCK_RESULT = {
  token: "heres a token",
  _id: "heres an id"
};
// const delay = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

// export const server = setupServer(
//   rest.post("http://localhost:3001/user", async (_req, res, ctx) => {
//     return res(ctx.json(SIGNUP_MOCK_RESULT));
//   })
// );