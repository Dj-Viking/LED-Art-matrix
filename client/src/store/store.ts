import { configureStore } from "@reduxjs/toolkit";
import { midiSlice } from "./midiSlice";
import { loggedInSlice } from "./loggedInSlice";
import { ledSlice } from "./ledSlice";
import { modalSlice } from "./modalSlice";
import { formSlice } from "./formSlice";
import { artScrollerSlice } from "./artScrollerSlice";
import { presetButtonsListSlice } from "./presetButtonListSlice";
import { GlobalState, MyRootState } from "../types";
import { useSelector } from "react-redux";
import { keyboardSlice } from "./keyboardSlice";

export const toolkitReducer = {
    midiState: midiSlice.reducer,
    loggedInState: loggedInSlice.reducer,
    keyboardState: keyboardSlice.reducer,
    ledState: ledSlice.reducer,
    modalState: modalSlice.reducer,
    formState: formSlice.reducer,
    artScrollerState: artScrollerSlice.reducer,
    presetButtonsListState: presetButtonsListSlice.reducer,
};

export const toolkitStore = configureStore({
    // silence the non-serializable errors because I just don't care about it.
    middleware(getDefaultMiddleware) {
        return getDefaultMiddleware({
            serializableCheck: false,
        });
    },
    reducer: toolkitReducer,
});

export function getGlobalState(selectorFn: typeof useSelector): GlobalState {
    const rootState = selectorFn((state: MyRootState) => state);

    let ret = {} as GlobalState;

    for (const sliceName of Object.keys(rootState)) {
        for (const sliceProp of Object.keys(rootState[sliceName])) {
            ret[sliceProp] = rootState[sliceName][sliceProp];
        }
    }

    return ret;
}

export type ToolkitRootState = ReturnType<typeof toolkitStore.getState>;

export type ToolkitDispatch = typeof toolkitStore.dispatch;
