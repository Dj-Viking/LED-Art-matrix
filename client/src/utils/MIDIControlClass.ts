/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React from "react";
import { artScrollerActions } from "../reducers/artScrollerSlice";
import { ledActions } from "../reducers/ledSlice";
import { determineDeviceControl, setAccess } from "../actions/midi-access-actions";
import { MIDIInputName, XONEK2_MIDI_CHANNEL_TABLE } from "../constants";
import { PresetButtonsList } from "./PresetButtonsListClass";
import { presetButtonsListActions } from "../reducers/presetButtonListSlice";

/**
 * @see https://www.w3.org/TR/webmidi/#idl-def-MIDIPort
 * interface MIDIPort : EventTarget {
    readonly    attribute DOMString               id;
    readonly    attribute DOMString?              manufacturer;
    readonly    attribute DOMString?              name;
    readonly    attribute MIDIPortType            type;
    readonly    attribute DOMString?              version;
    readonly    attribute MIDIPortDeviceState     state;
    readonly    attribute MIDIPortConnectionState connection;
                attribute EventHandler            onstatechange;
    Promise<MIDIPort> open ();
    Promise<MIDIPort> close ();
};
 */
interface MIDIPort {
    IODevice: MIDIInput | MIDIOutput;
    open: () => Promise<MIDIPort>;
    close: () => Promise<MIDIPort>;
}
interface TestMIDIConnectionEvent {
    isTrusted: boolean;
    bubbles: boolean;
    cancelBubble: boolean;
    cancelable: boolean;
    composed: boolean;
    target: MIDIInput;
}
interface MIDIConnectionEvent {
    isTrusted: boolean;
    bubbles: boolean;
    cancelBubble: boolean;
    cancelable: boolean;
    composed: boolean;
    currentTarget: MIDIAccessRecord;
    defaultPrevented: boolean;
    eventPhase: number;
    path: Array<any>;
    readonly PORT: MIDIPort;
    returnValue: boolean;
    srcElement: MIDIAccessRecord;
    target: MIDIAccessRecord;
    timeStamp: number;
    type: string | "statechange";
}
enum MIDIPortType {
    input = "input",
    output = "output",
}
enum MIDIPortConnectionState {
    open = "open",
    closed = "closed",
    pending = "pending",
}
enum MIDIPortDeviceState {
    disconnected = "disconnected",
    connected = "connected",
}

interface TestMIDIMessageEvent {
    isTrusted: boolean;
    bubbles: boolean;
    cancelBubble: boolean;
    composed: boolean;
    target: MIDIInput;
    data: [number, number, number];
}

interface MIDIMessageEvent {
    isTrusted: boolean;
    bubbles: boolean;
    cancelBubble: boolean;
    composed: boolean;
    currentTarget: MIDIInput;
    data: Uint8Array;
    defaultPrevented: boolean;
    eventPhase: number;
    path: Array<any>;
    returnValue: boolean;
    srcElement: MIDIInput;
    target: MIDIInput;
    timeStamp: number;
    type: "midimessage";
}

type onstatechangeHandler = null | ((event: MIDIConnectionEvent) => unknown);
interface MIDIInput {
    id: string;
    manufacturer: string;
    name: MIDIInputName;
    type: MIDIPortType.input;
    version: string;
    state: MIDIPortDeviceState;
    connection: MIDIPortConnectionState | string;
    onstatechange: undefined | onstatechangeHandler;
    onmidimessage: undefined | null | ((event: MIDIMessageEvent) => unknown);
}
interface MIDIOutput {
    connection: MIDIPortConnectionState;
    id: string;
    manufacturer: string;
    name: string;
    type: MIDIPortType.output;
    state: MIDIPortDeviceState;
    version: string;
    onstatechange: undefined | onstatechangeHandler;
    onmidimessage: undefined | null | ((event: MIDIMessageEvent) => unknown);
}

type TestMIDIAccessRecord = null | {
    readonly inputs: Map<MIDIInput["id"], MIDIInput>;
    readonly outputs: Map<MIDIOutput["id"], MIDIOutput>;
    onstatechange: null | ((event: TestMIDIConnectionEvent) => unknown);
    readonly sysexEnabled: boolean;
};
interface MIDIAccessRecord {
    readonly inputs: Map<MIDIInput["id"], MIDIInput>;
    readonly outputs: Map<MIDIOutput["id"], MIDIOutput>;
    onstatechange: onstatechangeHandler;
    readonly sysexEnabled: boolean;
}

interface IMIDIController {
    inputs?: Array<MIDIInput>;
    inputs_size: number;
    outputs_size: number;
    outputs?: Array<MIDIOutput>;
    online?: boolean;
    getInstance: () => this;
}

class MIDIController implements IMIDIController {
    public access = {} as MIDIAccessRecord;
    public inputs = [] as Array<MIDIInput>;
    public inputs_size = 0;
    public outputs_size = 0;
    public outputs = [] as Array<MIDIOutput>;
    public online = false;

    public constructor(access: MIDIAccessRecord) {
        this.access = access;
        this.online = true;
        this.inputs_size = access!.inputs.size;
        this.outputs_size = access!.outputs.size;
        this._setInputs(access!.inputs);
        this._setOutputs(access!.outputs);
    }

    public static async requestMIDIAccess(): Promise<MIDIAccessRecord> {
        // @ts-ignore because for some reason in vscode
        // this method doesn't exist on the navigator I guess..
        // only supported in chrome mostly for now
        return window.navigator.requestMIDIAccess();
    }

    public getInstance(): this {
        return this;
    }

    public setOutputCbs(): this {
        // OUTPUT CBS SETTING HERE
        for (let j = 0; j < this.outputs!.length; j++) {
            this.outputs![j].onstatechange = function (_event: MIDIConnectionEvent) {
                // console.log("output onstatechange event", event);
            };
            this.outputs![j].onmidimessage = function (_event: MIDIMessageEvent) {
                // console.log("output midimessage event", event);
            };
        }
        return this;
    }

    /**
        * this following loop sort of is confirming my theory that this class can
         have ownership callbacks passed to it maybe?? 
         
        * not too sure if things are passed by memory values or references into the class constructors
         in JS
        this effectively sets connection "open" on the devices somehow?? to send/recieve with hardware
     */
    public setInputCbs(
        _onmidicb?: (event: MIDIMessageEvent) => unknown,
        _onstatechangecb?: (event: MIDIConnectionEvent) => unknown
    ): this {
        for (let j = 0; j < this.inputs!.length; j++) {
            this.inputs![j].onstatechange = _onstatechangecb;
            this.inputs![j].onmidimessage = _onmidicb;
        }
        return this;
    }

    private _setOutputs(outputs: Map<string, MIDIOutput>): void {
        if (outputs.size > 0) {
            const MIDI_OUTPUT_LIST_SIZE = outputs.size;
            const entries = outputs.entries();

            for (let i = 0; i < MIDI_OUTPUT_LIST_SIZE; i++) {
                this.outputs!.push(entries.next().value[1]);
            }
        }
    }

    private _setInputs(inputs: Map<string, MIDIInput>): void {
        if (inputs.size > 0) {
            const MIDI_INPUT_LIST_SIZE = inputs.size;
            const entries = inputs.entries();

            for (let i = 0; i < MIDI_INPUT_LIST_SIZE; i++) {
                this.inputs!.push(entries.next().value[1]);
            }
        }
    }

    public static stripNativeLabelFromMIDIInputName(name: string): MIDIInputName {
        return name.replace(/(\d-\s)/g, "") as MIDIInputName;
    }

    public static mapMIDIChannelToController(midi_event: MIDIMessageEvent): void {
        console.log("midi event", midi_event);
    }

    public static handleXONEK2MIDIMessage(
        midi_event: MIDIMessageEvent,
        _setChannel: (channel: number) => void,
        _setIntensity: (intensity: number) => void,
        _dispatchcb: React.Dispatch<any>,
        timeoutRef: React.MutableRefObject<NodeJS.Timeout>,
        _buttonIds: string[]
    ): void {
        clearTimeout(timeoutRef.current);
        const midi_intensity = midi_event.data[2];
        const midi_channel = midi_event.data[1];

        _setChannel(midi_channel);
        _setIntensity(midi_intensity);

        const is_fader = /fader/g.test(XONEK2_MIDI_CHANNEL_TABLE[midi_channel]);
        const is_knob = /knob/g.test(XONEK2_MIDI_CHANNEL_TABLE[midi_channel]);

        _dispatchcb(
            determineDeviceControl({
                usingFader: is_fader,
                usingKnob: is_knob,
            })
        );

        switch (XONEK2_MIDI_CHANNEL_TABLE[midi_channel]) {
            case "1_a_button":
                timeoutRef.current = setTimeout(() => {
                    // TODO: when redux toolkit is in - we set styling based on our access to the whole state tree
                    // and not what we are passing into this handler because passing state stuff in here
                    // while it is changing causes memory leaks
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.setActiveButton(_buttonIds[0]));

                        PresetButtonsList.setStyle(
                            _dispatchcb,
                            "rainbowTest",
                            midi_intensity.toString()
                        );
                    }
                }, 20);
                break;
            case "1_b_button":
                timeoutRef.current = setTimeout(() => {
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.setActiveButton(_buttonIds[1]));

                        PresetButtonsList.setStyle(_dispatchcb, "v2", midi_intensity.toString());
                    }
                }, 20);
                break;
            case "1_c_button":
                timeoutRef.current = setTimeout(() => {
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.setActiveButton(_buttonIds[2]));

                        PresetButtonsList.setStyle(_dispatchcb, "waves", midi_intensity.toString());
                    }
                }, 20);
                break;
            case "1_d_button":
                timeoutRef.current = setTimeout(() => {
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.setActiveButton(_buttonIds[3]));

                        PresetButtonsList.setStyle(
                            _dispatchcb,
                            "spiral",
                            midi_intensity.toString()
                        );
                    }
                }, 20);
                break;
            case "2_e_button":
                timeoutRef.current = setTimeout(() => {
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.setActiveButton(_buttonIds[4]));

                        PresetButtonsList.setStyle(
                            _dispatchcb,
                            "fourSpirals",
                            midi_intensity.toString()
                        );
                    }
                }, 20);
                break;
            case "2_f_button":
                timeoutRef.current = setTimeout(() => {
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.setActiveButton(_buttonIds[5]));

                        PresetButtonsList.setStyle(_dispatchcb, "dm5", midi_intensity.toString());
                    }
                }, 20);
                break;
            case "1_lower_button":
                timeoutRef.current = setTimeout(() => {
                    if (midi_intensity === 127) {
                        _dispatchcb(presetButtonsListActions.toggleMidiMode());
                    }
                }, 20);
                break;
            case "1_upper_knob":
                timeoutRef.current = setTimeout(() => {
                    artScrollerActions.setSlider({
                        control: "circleWidth",
                        value: midi_intensity.toString(),
                    });
                }, 20);
                break;
            case "1_middle_knob":
                timeoutRef.current = setTimeout(() => {
                    artScrollerActions.setSlider({
                        control: "vertPos",
                        value: midi_intensity.toString(),
                    });
                }, 20);
                break;
            case "1_lower_knob":
                timeoutRef.current = setTimeout(() => {
                    artScrollerActions.setSlider({
                        control: "hPos",
                        value: midi_intensity.toString(),
                    });
                }, 20);
                break;
            case "2_upper_knob":
                timeoutRef.current = setTimeout(() => {
                    artScrollerActions.setSlider({
                        control: "invert",
                        value: midi_intensity.toString(),
                    });
                }, 20);
                break;
            case "2_middle_knob":
                timeoutRef.current = setTimeout(() => {
                    _dispatchcb(
                        artScrollerActions.setSlider({
                            control: "animDuration",
                            value: midi_intensity <= 0 ? "1" : midi_intensity.toString(),
                        })
                    );
                }, 20);
                break;
            case "1_fader":
                timeoutRef.current = setTimeout(() => {
                    _dispatchcb(
                        ledActions.setAnimVarCoeff(
                            (midi_intensity === 0 ? "1" : midi_intensity * 2).toString()
                        )
                    );
                }, 10);
                break;
            default:
                break;
        }
    }

    public static async setupMIDI(
        dispatchcb: React.Dispatch<any>,
        _setChannel: (channel: number) => void,
        _setIntensity: (intensity: number) => void,
        timeoutRef: React.MutableRefObject<NodeJS.Timeout>,
        _buttonIds: string[],
        _midiEditMode: boolean,
        _access: MIDIAccessRecord
    ): Promise<MIDIController> {
        return new Promise<MIDIController>((resolve) => {
            _access.onstatechange = (_event: MIDIConnectionEvent): void => {
                const onstatechangeAccess = new MIDIController(_event.target).getInstance();

                const midicb = function (midi_event: MIDIMessageEvent): void {
                    console.log("midi edit mode", _midiEditMode);
                    if (_midiEditMode) {
                        return MIDIController.mapMIDIChannelToController(midi_event);
                    }
                    if (midi_event.currentTarget.name.includes("XONE:K2")) {
                        return MIDIController.handleXONEK2MIDIMessage(
                            midi_event,
                            _setChannel,
                            _setIntensity,
                            dispatchcb,
                            timeoutRef,
                            _buttonIds
                        );
                    }
                };

                const onstatechangecb = function (_connection_event: MIDIConnectionEvent): void {
                    // console.log("CONNECTION EVENT SET INPUT CB CALLBACK", _connection_event);
                };

                dispatchcb(setAccess(onstatechangeAccess, midicb, onstatechangecb));
                resolve(onstatechangeAccess);
            }; // end onstatechange def
        });
    }
}

export type {
    MIDIConnectionEvent,
    MIDIInput,
    MIDIOutput,
    MIDIPort,
    MIDIAccessRecord,
    onstatechangeHandler,
    MIDIMessageEvent,
    TestMIDIAccessRecord,
    TestMIDIMessageEvent,
    TestMIDIConnectionEvent,
};

export { MIDIController, MIDIPortConnectionState, MIDIPortDeviceState, MIDIPortType };
