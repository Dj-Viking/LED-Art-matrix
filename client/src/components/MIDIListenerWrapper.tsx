/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { ReactNode, useEffect, useRef, useState } from "react";
import { keyGen } from "../utils/keyGen";
import { MIDIConnectionEvent, MIDIController, MIDIInput, MIDIMessageEvent } from "../utils/MIDIControlClass";
import { useDispatch, useSelector } from "react-redux";
import { setAccess } from "../actions/midi-access-actions";
import { MyRootState } from "../types";
import { animVarCoeffChange } from "../actions/led-actions";
import { XONEK2_MIDI_CHANNEL_TABLE } from "../constants";
import { setAnimDuration, setCircleWidth, setHPos, setInvert, setVertPos } from "../actions/art-scroller-actions";
import IntensityBar from "./IntensityBar";

interface MIDIListenerWrapperProps {
    children?: ReactNode | ReactNode[]
}

const MIDIListenerWrapper: React.FC<MIDIListenerWrapperProps> = (): JSX.Element => {
    const dispatch = useDispatch();
    const accessState = useSelector((state: MyRootState) => state.accessRecordState); 
    const { figureOn } = useSelector((state: MyRootState) => state.artScrollerState); 
    const [ size, setSize ] = useState<number>(0);
    const [ intensity, setIntensity ] = useState<number>(0);
    // channel four is xone:k2's upper left most knob above the first fader
    const [ channel, setChannel ] = useState<number>(4);
    const filterTimeoutRef = useRef<NodeJS.Timeout>(setTimeout(() => void 0, 500));

    useEffect(() => {
        (async (): Promise<void> => {
            if ("navigator" in window) {
                // request access from browser
                const access = await MIDIController.requestMIDIAccess();
                dispatch(setAccess(new MIDIController(access).getInstance()));
                // set size of inputs to re-render component at this moment of time
                setSize(access.inputs.size);
                //at this moment the promise resolves with access if size changed at some point
                if (size > 0) {
                    dispatch(setAccess(new MIDIController(access).getInstance()));
                    // define onstatechange callback to not be a function to execute when state changes later
                    access.onstatechange = function (_event: MIDIConnectionEvent): void {
                        
                        const onstatechangeAccess = new MIDIController(_event.target).getInstance();

                        const midicb = function (midi_event: MIDIMessageEvent): void {
                            clearTimeout(filterTimeoutRef.current);
                            if (midi_event.currentTarget.name.includes("XONE:K2")) {

                                const midi_intensity = midi_event.data[2];
                                const midi_channel = midi_event.data[1];

                                setChannel(midi_channel);
                                setIntensity(midi_intensity);

                                // console.log("dump data", midi_event);
                                switch(XONEK2_MIDI_CHANNEL_TABLE[midi_channel]) {
                                    case "1_upper_knob":
                                        filterTimeoutRef.current = setTimeout(() => {
                                            dispatch(setCircleWidth(midi_intensity.toString()));
                                        }, 20);
                                    break;
                                    case "1_middle_knob":
                                        filterTimeoutRef.current = setTimeout(() => {
                                            dispatch(setVertPos(midi_intensity.toString()));
                                        }, 20); 
                                    break;
                                    case "1_lower_knob": 
                                        filterTimeoutRef.current = setTimeout(() => {
                                            dispatch(setHPos(midi_intensity.toString()));
                                        }, 20);
                                    break;
                                    case "2_upper_knob": 
                                        filterTimeoutRef.current = setTimeout(() => {
                                            dispatch(setInvert(midi_intensity.toString()));
                                        }, 20);
                                    break;
                                    case "2_middle_knob": 
                                        filterTimeoutRef.current = setTimeout(() => {
                                            dispatch(setAnimDuration(midi_intensity.toString()));
                                        }, 20);
                                    break;
                                    // RATHER USE KEYS INSTEAD OF MIDI CONTROLLER BUTTONS
                                    case "1_upper_button": 
                                        // filterTimeoutRef.current = setTimeout(() => {
                                        //     dispatch(setFigureOn(figureOn ? true : false));
                                        // }, 100);
                                        void 0;
                                    break;
                                    case "1_fader": 
                                        // filterTimeoutRef.current = setTimeout(() => {
                                        //     dispatch(setFigureOn(figureOn ? true : false));
                                        // }, 100);
                                        //slight debounce to help with limiting dispatch
                                        filterTimeoutRef.current = setTimeout(() => {
                                            dispatch(animVarCoeffChange((midi_intensity).toString()));
                                        }, 10);
                                    break;
                                    default: break;
                                }
                            }
                        };

                        const onstatechangecb = function(_connection_event: MIDIConnectionEvent): void {
                            // console.log("CONNECTION EVENT SET INPUT CB CALLBACK", connection_event);
                        };

                        dispatch(setAccess(onstatechangeAccess, midicb, onstatechangecb));
                    };// end onstatechange def
                } // endif size > 0 
                // accessState dead zone
            }// endif "navigator" in window
        })();
    }, [dispatch, accessState.inputs.length, size, figureOn]);

    return (
        <>
            {
                // @ts-ignore
                accessState.online ? accessState.inputs!.map((input: MIDIInput, i: number, _arr: Array<MIDIInput>) => {
                    if (input.name.includes("XONE:K2")) {
                        return (
                            <div key={keyGen()} style={{ display: "flex", flexDirection: "column" }}>
                                <h2 key={keyGen()}>MIDI Device {i + 1}</h2>
                                <div key={keyGen()} style={{ width: "50%", margin: "0 auto", border: input.state === "connected" ? "solid 1px green" : " solid 1px red" }}> 
                                    <p key={keyGen()} style={{marginBottom: "1em"}}>
                                        {input.name}
                                    </p>
                                    <span key={keyGen()}>
                                        
                                        Connection: { input.connection } 
                                        
                                        <div key={keyGen()} style={{ margin: "0 auto 0 auto", width: "40px", height: "0px", backgroundColor: input.connection === "connected" ? "black" : "black", borderRadius: "50%", border: "solid black 1px" }}></div> 
    
                                        <span>
                                            { 
                                                input.name.includes("XONE:K2") && (
                                                    <>
                                                        <p>
                                                            Intensity: {intensity} 

                                                            <IntensityBar intensity={intensity} />
                                                        </p>
                                                        <p>
                                                            Channel: { channel }
                                                        </p>
                                                    </>
                                                )
                                            }
                                        </span>
    
                                    </span>
                                    <p key={keyGen()}>{input.name}</p>
                                </div>
                            </div>
                        );
                    } else { //only want to display xone:k2 device
                        return null;
                    }
                }) : <p>MIDI OFFLINE</p>
            }
        </>
    );
};

export default MIDIListenerWrapper;