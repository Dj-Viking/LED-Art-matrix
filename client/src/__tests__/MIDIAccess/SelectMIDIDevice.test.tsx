/* eslint-disable testing-library/no-unnecessary-act */
/* eslint-disable @typescript-eslint/no-non-null-assertion */
//@ts-ignore
import React from "react";
import { combinedReducers } from "../../store";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { fireEvent, render, screen } from "@testing-library/react";
//@ts-ignore
import "@types/jest";
import "@testing-library/jest-dom";
import "@testing-library/jest-dom/extend-expect";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { MIDIAccessRecord } from "../../utils/MIDIControlClass";
import MIDIListenerWrapper from "../../components/MIDIListenerWrapper";
import { act } from "react-dom/test-utils";
import { TestService } from "../../utils/TestServiceClass";
import { MOCK_MIDI_ACCESS_RECORD } from "../../utils/mocks";

const store = createStore(
    combinedReducers,
    // @ts-expect-error this will exist in the browser
    window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);

// @ts-ignore need to implement a fake version of this for the jest test as expected
// did not have this method implemented by default during the test
global.navigator.requestMIDIAccess = async function (): Promise<MIDIAccessRecord> {
    return Promise.resolve(MOCK_MIDI_ACCESS_RECORD);
};

TestService.createMockMIDIControllerClass();

describe("test selecting midi device toggles which device shows", () => {
    it("select device from dropdown menu", async () => {
        const history = createMemoryHistory();

        render(
            <Provider store={store}>
                <Router history={history}>
                    <MIDIListenerWrapper />
                </Router>
            </Provider>
        );

        const select = {
            dropdown: (await screen.findByTestId("midi-select")) as HTMLElement,
            options: (await screen.findAllByTestId("select-option")) as HTMLElement[],
        };

        expect(select.dropdown).toBeInTheDocument();
        expect(select.dropdown).toHaveValue("Select A Connected Device");

        expect(select.options).toHaveLength(4);

        act(() => {
            fireEvent.change(select.dropdown, { target: { value: "XONE:K2 MIDI" } });
        });

        expect(select.dropdown).toHaveValue("XONE:K2 MIDI");
    });
});
