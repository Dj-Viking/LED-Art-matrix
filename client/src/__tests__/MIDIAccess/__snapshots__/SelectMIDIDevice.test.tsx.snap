// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test selecting midi device toggles which device shows select device from dropdown menu 1`] = `
<DocumentFragment>
  <h2>
    MIDI Devices
  </h2>
  <div
    style="display: flex; flex-direction: column; justify-content: center;"
  >
    <div
      style="display: flex; align-items: center; justify-content: center;"
    >
      <select
        data-testid="midi-select"
        style="background-color: black;"
      >
        <option
          data-testid="select-option"
          disabled=""
          value="Select A Connected Device"
        >
          Select A Connected Device
        </option>
        <option
          data-testid="select-option"
          value="XONE:K2 MIDI"
        >
          XONE:K2 MIDI
        </option>
        <option
          data-testid="select-option"
          value="XONE:K2 MIDI"
        >
          XONE:K2 MIDI
        </option>
        <option
          data-testid="select-option"
          value="XONE:K2 MIDI"
        >
          XONE:K2 MIDI
        </option>
      </select>
    </div>
    <div
      style="position: relative; width: 50%; margin: 0px auto; border: 1px solid green;"
    >
      <p>
        XONE:K2 MIDI
      </p>
      <div
        class="sc-bczRLJ gGpyYT"
      />
      <span>
        Intensity: 0
      </span>
      <div
        style="margin-top: 1em; display: flex; flex-direction: column; justify-content: center; align-items: center;"
      >
        <progress
          max="127"
          style="width: 50%;"
          value="0"
        />
      </div>
      <div
        class="sc-hKMtZM fCrJUW"
      >
        <span>
          Channel: 16
        </span>
        <p
          style="margin: 0px;"
        >
          1_fader
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
